returnn/datasets/distrib_files.py

```diff
@@ -245,8 +266,32 @@ def init_seq_order(self, epoch: Optional[int] = None, seq_list=None, seq_order=N
         self._lazy_init_file_sizes()
         self._lazy_init_file_cache_proc()
 
-        # Cleanup and fill _files_order_cache.
         full_epoch_0idx = (epoch - 1) // self.partition_epoch
+
+        # Shard files across GPU workers
+        for k in list(self._files_for_this_worker.keys()):
+            if k < full_epoch_0idx:
+                del self._files_for_this_worker[k]
+        for ep_ in range(epoch, epoch + self.preload_next_n_sub_epochs + 1):
+            full_epoch_0idx_ = (ep_ - 1) // self.partition_epoch
+            if full_epoch_0idx_ in self._files_for_this_worker:
+                continue
+            if self.seq_ordering == "default":
+                files = self.files
+            elif self.seq_ordering == "random":
+                # do not use `self._get_random_seed_for_epoch` to exclude `random_seed_offset`
+                # to ensure an equal distribution in every worker
+                random_generator = numpy.random.RandomState(full_epoch_0idx_)
```

Comment:
Is this the right way to ensure equal distribution across all workers, does this need to take any more info into account or is the full epoch number enough here?

Reply:
Full epoch is enough. But I don't like to just take it as-is as a random seed. Multiply it with some prime number and add some prime number. But also, just use `_get_random_seed_for_epoch` here.

Reply:
Is there a good reason for that?

Reply:
I just want to have a different random state at different places where we use the epoch for seeding. If we just use `RandomState(epoch)` in multiple places throughout RETURNN (e.g. for shuffling seqs, for data augmentation, for random param init, whatever), it means all those RNGs start from the same state. This could have some weird effects. I like that all RNGs start from different seeds, by doing those adjustments on the seed. It's a bit similar reasoning like [salting for cryptography](https://en.wikipedia.org/wiki/Salt_(cryptography)).

Reply:
> Multiply it with some prime number and add some prime number.

now the same prime number is chosen for both multiplier and summand which makes 
```
full_epoch_0idx_ * 127 + 127
```
equivalent to
```
(full_epoch_0idx_ + 1) * 127
```

which still works, I guess, but was this how it was intended?

Reply:
No, I did not intend to use the same prime for both multiplier and summand. I think this should be a different prime. Also, I would use a much larger prime. Also, I like primes of the form 2^n - 1, but probably that doesn't matter.

As an example, see how we seed the PyTorch random number generator:
```python
random_seed = self.config.int("random_seed", 42)
random_seed = (epoch * 193939 + step * 19937 + random_seed * 27644437 + 479001599) % (2**31)
rf.set_random_seed(random_seed)
```

Reply:
> Also, I like primes of the form 2^n - 1

to be fair, 127 is `2^7 - 1` :P